[NETBEANS-1074] Module review j2ee.dd

juneau001 opened the pull request:
- Added license to binaries-list
- Modified external dependency licenses accordingly
- Added license header to files, where missing
- Checked against Rat
- Scanned sources...no other changes needed.  Left data classes un-licensed'

matthiasblaesing commented on enterprise/j2ee.dd/external/binaries-list:
```diff
@@ -1,3 +1,19 @@
+# under the License.
 D2F2B34B7D0A13256D491B14590ED23CA0E49B46 javaee-api-5.jar
 2801A0BB60C19040FDA8C4801963911255774718 j2ee.dd-glassfish-resources.zip
 12763FC339B86791D848078A479793C3FF815982 j2ee.dd-w3c-resources.zip
```
It would be great if you could check if you can remove the use of the custom download server. The javaee-api should be available on maven central, for the other two resources, you could see if you can adapt the methods Reema and Sarvesh  applied:

https://github.com/apache/incubator-netbeans/pull/935
https://github.com/apache/incubator-netbeans/pull/877

If you feel uncomfortable with that, please indicate so, that I'll take a look at it.

juneau001 replied:
@matthiasblaesing To resolve the ZIP external dependencies, can I find the files contained within the ZIP and add them to the project...of course adding the appropriate license?  For instance, with the w3c-resources, there are three files within that archive.  Can I simply copy those files into the appropriate location within the j2ee.dd module structure, or do I need to obtain those files from maven central somewhere?  Thanks

matthiasblaesing replied:
It depends. For XSDs + DTDs it can be argued, that they are used at runtime and thus need to be included in their source form. We are currently integrating them if possible (apache legal faq describes this as Category B). The referenced PRs take that road.

For the javaee-api-5 dependency, this should be switch to the maven dependency `javaee:javaee-api:5`

juneau001 commented on enterprise/j2ee.dd/licenseinfo.xml:
```diff
@@ -0,0 +1,30 @@
+    <fileset>
+        <file>doc/dd_api_schema.jpg</file>    
+        <license ref="Apache-2.0-ASF" />
+        <comment type="TEMPLATE_MINIMAL_IP"/>
```
Thanks!  Just curious...is there a list somewhere of the different values for comment type that I could use for reference?  I'll make the change.

matthiasblaesing replied:
In the source :-) - Seriously I set up a WikiPage to describe it: https://cwiki.apache.org/confluence/display/NETBEANS/Legal+FAQ

juneau001 commented:
Unsure as to why the Travis CI build failed, as I ran and passed all tests on my machine.  @matthiasblaesing any ideas to point me in the right direction to repair the issue?  Thanks

matthiasblaesing commented:
@juneau001 I'll start with your build error - lets start how to see it:
- Click on "details" behind the travis check - you now get to the travis-ci web interface
- Click on the failing job - you'll see the normal output (incomplete)
- Click on "Raw log" - you'll get to: https://api.travis-ci.org/v3/job/441151558/log.txt
- Now look for "Build failed" and you'll get:

```
travis_time:start:00449ade
[0K$ if [ "x$LICENSE" == "xtrue" ]; then ant -quiet verify-libs-and-licenses -Dverify-libs-and-licenses.haltonfailure=true; fi
     [echo] Bootstrapping NetBeans-specific Ant extensions...
     [echo] Downloading external binaries (*/external/ directories) for cluster.config=release...
[verifylibsandlicenses] Warning: org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser: Property 'http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit' is not recognized.
[verifylibsandlicenses] Compiler warnings:
[verifylibsandlicenses]   WARNING:  'org.apache.xerces.jaxp.SAXParserImpl: Property 'http://javax.xml.XMLConstants/property/accessExternalDTD' is not recognized.'

BUILD FAILED
/home/travis/build/apache/incubator-netbeans/nbbuild/build.xml:1532: Failed VerifyLibsAndLicenses test(s):
Some licenseinfo.xml files failed verification:
enterprise/j2ee.dd/licenseinfo.xml referenced file not found 'src/org/netbeans/modules/j2ee/dd/impl/resources/java_web_services_client_1_1.xsd'
```

You can get the full check on your system by running:
```
ant -Dcluster.config=enterprise verify-libs-and-licenses
```
from the root of the repository. The output can be found in: `nbbuild/build/verifylibsandlicenses.xml`. Look only for failures referencing the module you are working on.

Long story short: Your `licenseinfo.xml` references a file, that is not present.

The build failure is IMHO minor - a bigger problem is, that your `licenseinfo.xml` makes invalid license claims.

- You need to create one fileset per license/comment tuple and I think this is the main problem.
- You imported the XSDs. Most, if not all, DTDs/XSDs Oracle/Sun created for Open Source projects are dual-licenses as GPLv2-CPE/CDDL. For the Apache Netbeans Project we elect the CDDL as the license. Yet the file declares: `<license ref="Apache-2.0-ASF" />`. You need to check the individual files (they all should contain a license statement), but most probably are GPLv2-CPE/CDDL-1.1 and thus a correct entry would be `<license ref="CDDL-1.1" />` for these.
- you declare `<comment type="COMMENT_UNSUPPORTED"/>` that is correct for the `doc/dd_api_schema.jpg` file. The comment should give a reasoning why the file can/needs to be included. For the schemas we include them because we use the schemas at runtime for XML validation and code completion. The CDDL can be included (https://www.apache.org/legal/resolved.html#category-b, last paragraph). The prepared comment for that is "CATEGORY-B". If you read the FAQ entry the name becomes clear :-)

So please:
- check the files if valid license headers/license info is present, if not please try to find a good copy (for the standard a good source is:https://search.maven.org/artifact/org.glassfish.main.appclient/gf-client-module/5.0/glassfish-jar, for W3C files please look on the W3C site)
- If the file is from the W3C, the license is most probably `W3C-2` (the current version of the W3C software license). Please see `enterprise/web.core/src/org/netbeans/modules/web/taglib/resources/xml.xsd` and `enterprise/web.core/licenseinfo.xml` for a sample. I added the license header to the file and documented that inline.
- The JavaEE XSDs normally have valid headers (see discussion about license above). It just needs to be checked, that it is ok.
- Group the files by license and comment and add `fileset` entries for each group of file to `licenseinfo.xml`

juneau001 commented:
Thanks for the detailed information @matthiasblaesing  I was not aware of the requirements for the licenseinfo.xml file, so this is great information!  I will make the requested changes and repair the issue that is causing the failure.  I certainly appreciate it!

matthiasblaesing commented:
Thank you for the updated PR. In general this looks good. I noticed this:
- `web-app_2_2.dtd` is declared as originate from the W3C. I doubt that. Judging from the other descriptors, I would replace it with a copy from here: `https://www.jcp.org/j2ee/dtds/web-app_2_2.dtd` and adjust the `licenseinfo.xml` accordingly
- `web-app_2_3.dtd` looks strange (the license is unconventional) - this copy hold a cleaner license: https://www.jcp.org/dtd/web-app_2_3.dtd. 

Both files can also be found in the following glassfish artifact:
https://search.maven.org/artifact/org.glassfish.main.appclient/gf-client-module/5.0/glassfish-jar
Given that oracle was the distributor/developer of glassfish this would be a safe source. With these files updated, I would merge.

juneau001 commented:
@matthiasblaesing I appreciate the assistance.  I've updated the files using those to which you had supplied links.  I've also modified licenseinfo.xml accordingly.  Hope this does the trick...thanks again!

matthiasblaesing commented:
Thanks for the update. I have two remaining issues:

- the licenses of the javaee-api-5.jar is not ALv2 (see the `META-INF/LICENSE.TXT` inf the file)
- the two license file `glassfish-license.txt` and `w3c-license.txt` don't belong into the `external` directory, as the corresponding files were removed. The run of `ant -Dcluster.config=enterprise verify-libs-and-licenses` should report that (the travis-ci runs currently only cover a subset of the codebase).

I also added a `.gitignore` to the module directory, so that the generated files from the XSDs are not reported as "new" files for git.

I squashed your commits into one and added the changes described above as a second commit:
https://github.com/matthiasblaesing/incubator-netbeans/commits/github-pr-916

The update:
https://github.com/matthiasblaesing/incubator-netbeans/commit/f1c806ff5cdad04f2a16dd518d474ca1cc03919e

It would be great if you could have a look. If you agree, I would squash these two commits together, leave you as the author and merge that.

juneau001 commented:
Thanks for all of your assistance @matthiasblaesing ... your update looks great.  I agree that these updates should be merged.

matthiasblaesing commented:
Merged as discussed - thank you for working on this.
